lib/Twig/Extensions/Node/Trans.php

```diff
@@ -51,17 +65,37 @@ public function compile(Twig_Compiler $compiler)
         if ($vars) {
             $compiler
                 ->write('echo strtr('.$function.'(')
-                ->subcompile($msg)
             ;
-
-            if (null !== $this->getNode('plural')) {
+            if(!is_null($context) && !is_null($plural)){
+                $compiler
+                    ->string($context)
+                    ->raw(', ')
+                    ->subcompile($msg)
+                    ->raw(', ')
+                    ->subcompile($msg1)
+                    ->raw(', abs(')
+                    ->subcompile($this->getNode('count'))
+                    ->raw(')')
+                ;
+            }elseif(!is_null($context)){
+                $compiler
+                    ->string($context)
+                    ->raw(', ')
+                    ->subcompile($msg)
+                ;
+            }elseif(!is_null($plural)){
                 $compiler
+                    ->subcompile($msg)
                     ->raw(', ')
                     ->subcompile($msg1)
                     ->raw(', abs(')
                     ->subcompile($this->getNode('count'))
                     ->raw(')')
```

Comment:
you could avoid lots of duplication here, as the context is always first when it is there, then followed by ``$msg``, then followed by the pluralization arguments.
You only need to add a check for the context before the ``->subcompile($msg)`` call